test: add common.mustSucceed

tniessen opened the pull request:
Our increasingly large collection of tests makes frequent use of this pattern:

```javascript
someFunction(/* some args */, common.mustCall((err, /* some results */) => {
  assert.ifError(err);
  // Something else
});
```

This PR changes those to this pattern:

```javascript
someFunction(/* some args */, common.mustSucceed((/* some results */) => {
  // Something else
});
```

This simple change allows us to remove more than 300 lines of code from tests. Another aspect is that many tests really should be using `mustSucceed()` instead of `mustCall()`, because `mustCall()` silently ignores errors passed to it. I changed the crypto tests manually where I found such cases, and a few other test files, but there's probably dozens or hundreds of such cases in other test files.

---

To make this easier to review, I separated the changes into multiple commits.

1. The first commit introduces the new API.
2. The next three commits apply regular expressions for automatic conversions from `mustCall` to `mustSucceed`.
   1. The first replaces all occurrences of `mustCall\([ \n]*(\(erro?r?\)[ \n]*=>[ \n]*\{?|function[ \n]*\(erro?r?\)[ \n]*\{)?[ \n]*assert\.ifError(\(erro?r?\))?;?[ \n]*\}?[ \n]*\)` with `mustSucceed()`. These are trivial cases, e.g, `mustCall(assert.ifError)`, `mustCall((err) => assert.ifError(err))`, `mustCall(function(err) { assert.ifError(err); })` etc.
   2. The second replaces all occurrences of `mustCall\([\n ]*\(erro?r?(,[\n ]*([^\)]*))?\)([\n ]*)=>([\n ]*)\{[\n ]+assert\.ifError\(erro?r?\);\n` with `mustSucceed(($2)$3=>$4{\n`. These are arrow functions that take an error argument and begin with `assert.ifError`.
   3. The third replaces all occurrences of `mustCall\([\n ]*function[ \n]*\(erro?r?(, *([^\)]*))?\)([ \t\n]*)\{[\n ]+assert\.ifError\(erro?r?\);\n` with `mustSucceed(function($2)$3{\n`. These are regular functions that take an error argument and begin with `assert.ifError`.
3. The fifth commit converts functions containing statements such as `assert.ok(!err)` to `mustSucceed`, where possible.
4. The sixth commit improves formatting. While all previous commits pass tests and linter, the regular expressions leave some unnecessary empty lines. While at it, I also converted some functions to arrow functions (only in lines that I changed anyway).
5. The seventh commit replaces the weaker `mustCall` with `mustSucceed` where it seems to make sense.

---

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

<!--
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

Trott commented:
My immediate reaction is that I prefer the explicitness and clarity of the current pattern. `common.mustSucceed()` seems be just a bit too magical for my tastes, and both brevity and lack of repetition in tests are overrated. (Lack of repetition is a virtue almost everywhere else, though.)

But that's my reaction without looking much and I have to admit you make some good points in your explanation. I'll take a closer look later, but for now: /ping @nodejs/testing

Trott commented:
> My immediate reaction is that I prefer the explicitness and clarity of the current pattern.

Having spent a little more time with this, I think I"m a convert. This seems like a good idea to me.

Trott reviewed (dismissed):
The new function needs to be documented in `test/common/README.md`. Also, might be worthwhile looking for `common.mustCall()` instances in our internal docs that need to be changed. I'm thinking specifically of `doc/guides/writing-tests.md` but there may be others.

lpinca commented:
> My immediate reaction is that I prefer the explicitness and clarity of the current pattern.

I agree, and I think this does not help much with cases where `mustCall()` ignores the error. Even if this is added the test writer could use `mustCall()` instead of `mustSucceed()`. If reviewers notice that the error is ignored they can ask the test writer to use `mustSucceed()` but at this point, isn't it easier and clearer if they ask the test writer to use `assert.ifError(err);`.

It's test writer responsibility to make sure errors are not ignored with or without this change.

tniessen commented:
> My immediate reaction is that I prefer the explicitness and clarity of the current pattern. `common.mustSucceed()` seems be just a bit too magical for my tastes, and both brevity and lack of repetition in tests are overrated.

Personally, I think that the `callback(err)` pattern is such a natural part of Node.js that we should not treat `err` as a regular argument. In most tests, its value has absolutely no use, except that the test should fail if it is not `null` or `undefined` (and many tests do not implement that). `mustSucceed` would simply reinforce the pattern of seeing `err` as an exception, not a regular argument.

We have something remotely similar for `Promise` APIs, [`assert.doesNotReject`](https://nodejs.org/api/assert.html#assert_assert_doesnotreject_asyncfn_error_message), and we came to the conclusion that it isn't helpful. For callback APIs, we have absolutely no test utility, even though it would be helpful.

> I think this does not help much with cases where `mustCall()` ignores the error. Even if this is added the test writer could use `mustCall()` instead of `mustSucceed()`.

From my perspective, simply having `mustSucceed()` and beginning to use it would have a chance of counteracting the fact that many tests use `mustCall()` incorrectly.

> If reviewers notice that the error is ignored they can ask the test writer to use `mustSucceed()` but at this point, isn't it easier and clearer if they ask the test writer to use `assert.ifError(err);`?

I actually started this because I noticed a lot of tests using `common.mustCall()` without an error handler as a callback that receives an `err`. Personally, I think having `mustSucceed()` as a shorthand for `mustCall(assert.ifError)` is neither less readable nor understandable. From there, extending it to functions that do their own validation seemed to make sense.

tniessen commented:
Kindly asking @jasnell, @BridgeAR, @cjihrig, @addaleax for input due to their experience with these particular tests.

```
$ git rev-parse HEAD
1adb4cd0b18fa9dac257def9dbeabe7e687be1b0
$ git rev-parse master
bb9117ee9f8fb42ac8efd1dab30b3faee5463587
$ git log --format='%aN' master -- $(git diff master --name-only) | sort | uniq -c | sort -r | head -n5
    138 Rich Trott
     75 James M Snell
     72 Ruben Bridgewater
     51 Colin Ihrig
     47 Anna Henningsen
```

addaleax commented:
My initial thought would be that this only really makes sense if it is enforced through a linter rule, because I would expect that this pattern does look surprising to people who are new to our test suite (the same obviously goes for most patterns that are specific to it).

tniessen commented:
> My initial thought would be that this only really makes sense if it is enforced through a linter rule

I think we could use a linter rule to prevent `common.mustCall((err) => assert.ifError(err))` etc., but not much beyond that (so no wrong use of `mustCall´). Is that what you mean?

addaleax commented:
@tniessen Yeah, any `common.mustCall()` where the function body contains `assert.ifError(err)`

tniessen commented:
> Yeah, any `common.mustCall()` where the function body contains `assert.ifError(err)`

@addaleax I added the rule and fixed the resulting linter problems. However, I restricted it to only apply if the function body begins with `assert.ifError`, because some tests are allowing some errors in the callback, and only rejecting certain errors, e.g., like this:

```js
something(common.mustCall((err) => {
  if (err && err.code === 'ERR_SOMETHING') {
    // Do something
    return;
  }
  assert.ifError(err);
});
```

tniessen commented:
This has been open for more than three weeks and has neither been approved nor rejected nor otherwise reviewed. Ping @nodejs/testing (hopefully that's a good choice?).

Trott commented:
> This has been open for more than three weeks and has neither been approved nor rejected nor otherwise reviewed. Ping @nodejs/testing (hopefully that's a good choice?).

I go back and forth between thinking it's a good idea and might help catch some bugs in our tests...then thinking it may not be worth the churn/additional API...and then thinking that it doesn't matter that much either way because we can do whatever we want with the tests and then change it if we don't like it after a month or whatever.

Let's give this another 3 days or so and if no one else either approves or objects, I'll approve it and you can rebase and land. I want to give people with stronger opinions (and who are putting more thought into it) a chance to weigh in. But I **think** I like this change.

tniessen commented:
@Trott It's been another two weeks, I will rebase this, and you (or anyone else) can either approve or close this. It doesn't seem like anyone cares either way.

tniessen commented:
Ping @Trott. Sorry for pinging you again, I just don't want this to be forgotten for another few weeks and require more rebases. Feel free to ignore or close if you don't think approving it is the right thing to do.

ruyadorno commented:
This is awesome @tniessen! Love it that you provided the eslint rules and the PR even catchs a bunch of scenarios of `mustCall()` being used without asserting the error.

I'm a big proponent of less abstractions in tests but all things weighted this is def one worth having 👍

aduh95 commented:
Landed in 30fb4a015db2

tniessen commented:
Thank you for reviewing @ruyadorno and @Trott, and thank you for landing @aduh95.
